File: egress_hub/code/cache_account_updater/SignalDecoder/SignalDecoder.cpp
#include "SignalDecoder.h"

#include "AppConfig.h"

using namespace digital::exchange::sbe::hub::inner;

std::unique_ptr<Signal> SignalDecoder::decodeSignal(char *buffer, int64_t offset, int64_t length)
{
    if(length < signalDecoder.SBE_BLOCK_LENGTH)
    {
        return nullptr;
    }

    messageHeaderDecoder.wrap(buffer, offset, messageHeaderDecoder.actingVersion(), length);

    if(messageHeaderDecoder.templateId() != signalDecoder.SBE_TEMPLATE_ID)
    {
        return nullptr;
    }

    signalDecoder.wrapForDecode(
        buffer,
        offset + messageHeaderDecoder.encodedLength(),
        messageHeaderDecoder.blockLength(),
        messageHeaderDecoder.actingVersion(),
        length
    );

    if(!(signalDecoder.hubServiceIdMask() & AppConfig::serviceId))
    {
        return nullptr;
    }

    return std::make_unique<Signal>(
        signalDecoder.hubServiceIdMask(),
        signalDecoder.position(),
        transformSignal(signalDecoder.signal())
    );
}

SignalCommandEnum SignalDecoder::transformSignal(const SignalEnum::Value signal)
{
    switch (signal)
    {
    case CONNECT:
        return SignalCommandEnum::CONNECT;

    case DISCONNECT:
        return SignalCommandEnum::DISCONNECT;

    case DO_WORK:
        return SignalCommandEnum::DO_WORK;
    
    default:
        return SignalCommandEnum::NULL_VAL;
    }
}

File: egress_hub/code/me_egress_poller/src/main/java/digital/exchange/egress/hub/me/egress/poller/hub/SignalDecoder.java
package digital.exchange.egress.hub.me.egress.poller.hub;

import digital.exchange.egress.hub.me.egress.poller.utils.AppConfig;
import digital.exchange.sbe.hub.inner.SignalEnumFromHubNodeDecoder;
import digital.exchange.sbe.hub.inner.MessageHeaderDecoder;
import org.agrona.DirectBuffer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SignalDecoder
{
    private static final Logger LOGGER = LoggerFactory.getLogger(SignalDecoder.class);

    private static final MessageHeaderDecoder messageHeaderDecoder = new MessageHeaderDecoder();
    private static final SignalEnumFromHubNodeDecoder signalDecoder = new SignalEnumFromHubNodeDecoder();

    public static SignalEnumFromHubNodeDecoder decodeSignal(DirectBuffer buffer, int offset, int length)
    {
        if(length < SignalEnumFromHubNodeDecoder.BLOCK_LENGTH)
        {
            LOGGER.info("Message length too small. skipping");
            return null;
        }

        signalDecoder.wrapAndApplyHeader(buffer, offset, messageHeaderDecoder);

        if(messageHeaderDecoder.templateId() != SignalEnumFromHubNodeDecoder.TEMPLATE_ID)
        {
            LOGGER.info("Not a command message. skipping");
            return null;
        }

        if((signalDecoder.hubServiceIdMask() & AppConfig.serviceId) == 0)
        {
            return null;
        }

        return signalDecoder;
    }
}
